### ISSUE TITLE
ERROR:root:Exception deserializing message!

### ISSUE THREAD

**Issue by zurk** [bblfsh/client-python#89]
How to reproduce:
```
wget https://raw.githubusercontent.com/lumoslabs/aleph/master/vendor/assets/javascripts/ace/mode-pgsql.js
```
in python:
```python
import bblfsh

client = bblfsh.BblfshClient("0.0.0.0:9432")
client.parse("mode-pgsql.js")
```

And you will get:
```
ERROR:root:Exception deserializing message!
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/grpc/_common.py", line 87, in _transform
    return transformer(message)
google.protobuf.message.DecodeError: Error parsing message
Traceback (most recent call last):
  File "/home/k/.local/lib/python3.5/site-packages/IPython/core/interactiveshell.py", line 2910, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-33-6f0fa647fcd6>", line 1, in <module>
    resp = bf.parse("/home/k/sourced/workdir/mode-pgsql.js")
  File "/usr/local/lib/python3.5/dist-packages/bblfsh/client.py", line 74, in parse
    return self._stub.Parse(request, timeout=timeout)
  File "/usr/local/lib/python3.5/dist-packages/grpc/_channel.py", line 487, in call
    return _end_unary_response_blocking(state, call, False, deadline)
  File "/usr/local/lib/python3.5/dist-packages/grpc/_channel.py", line 437, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.INTERNAL, Exception deserializing response!)>
```
versions:
* python-client: 2.9.6
* bblfsh container: 
```
REPOSITORY                       TAG                 IMAGE ID                                                                  CREATED             SIZE
bblfsh/bblfshd                   latest              sha256:fb5eb6936c67c20f451daddc09a4648cc26da978041af36f9f70b246f71d804d   13 days ago         173MB
```  
* javascript driver:
```
| javascript | docker://bblfsh/javascript-driver:latest | dev-adcd1b4 | beta      | 11 days  | alpine | 1.9 | 8.9.3-r0    |
```

I am not sure is it relatet to client-python or to javascript driver...

**Comment by juanjux** [comment]
Looks like a problem with the javascript driver returning a broken response, I'll confirm in a couple minutes.

**Comment by juanjux** [comment]
After some tests, looks like it could be a problem with the Python client, as reported. I've run this without bblfshd (to discard problems with it), just with the javascript driver and it works with the go client but not with the python one.

So the problem must be in the Python client, but it looks like it fails before it arrives to any of our code, just in the gRPC handling. I'll try to go back to grpcio 1.9.1 since it looks like the upstream 1.10.0 dissapeared again.

**Comment by zurk** [comment]
yeah, I also find some other files in different languages where python client fails. So, seems to be driver unrelated problem.

**Comment by smola** [comment]
I'm experiencing the same bug with [this file](https://github.com/go-lang-plugin-org/go-lang-idea-plugin/blob/e73dab44213ffa76b3d6a853aecb109929c3e2b5/gen/com/goide/GoTypes.java) (Java). Other files work correctly. The problem seems to be only with deserialization from Python.

bblfsh 2.9.6 in Python here too.

**Comment by smola** [comment]
I found this problem for the first time while using src-d/ml on the PGA dataset. In that case, the exception was raised when deserializing a node with `Node.FromString`, because in that case, we first load the raw protobuf from bblfsh, and we parse it at a later stage.

Here's a gist with an example Java file that exhibits this bug, as well as the raw proto message that we got and cannot deserialize:
https://gist.github.com/smola/e1f52b1070d5c6d346624cfdeadc97da

**Comment by juanjux** [comment]
Thanks @smola, I'll take a second look at this with the new data.

**Comment by juanjux** [comment]
Still not conclusive, but I'm almost sure is a Python grpc problem.

I've created another gist:

https://gist.github.com/juanjux/396fe25f9bc7d1c3fec8a8d4c47bcb28

As you can see, the java file is much reduced. Any change on it (removing the outer class, removing one of the `else if` or changing the `return` value) will make it work. 

I've attached to that gist the proto file that crashes and one that works after a minimal change (removing one of the `else if`s).

Debugging the code it finally fails in `grpc/_common.py` in the `_transform` function that takes a transformer. That parameter in my computer is a native function, so I can't easily debug without recompiling grpc from sources (very ugly task, from experience.)

I'll try with older versions of grpc and if nothing works I'll add an issue upstream.

**Comment by juanjux** [comment]
https://github.com/grpc/grpc/issues/14847

**Comment by juanjux** [comment]
I've been fighting with this a little more. As suggested in the issue linked in the previous message, I exported (marshalled) a ParseResponse from Go and tried to import (ParseFromString) from Python, without gRPC involved. It fails again. So it seems like the problem is or a protobuf bug (the file is not huge, 20kb, so I don't think we're hitting any limits).

Anyway, there is a workaround: setting the `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION` environment variable to `python` make it works (but is probably a lot slower). I'll open a new issue in the protobuf repository and meanwhile I'll add an ugly workaround setting that variable when we receive that error and resetting after it. If you're loading the exported files outside the client but with Python (@smola) you should probably do the same.

**Comment by juanjux** [comment]
Nope, the workaround doesn't work, the environment variable must be set before the initial import (even reloading the grpc and protobuf modules after setting the variable doesn't have effect).

**Comment by juanjux** [comment]
google/protobuf/issues/4484

**Comment by zurk** [comment]
Any updates about the issue?

**Comment by vmarkovtsev** [comment]
We keep hitting this indeed.

**Comment by zurk** [comment]
the same way to reproduce works for bblfshd `v2.10.0` and js drivers `v1.2.0` and `v2.5.0`

**Comment by juanjux** [comment]
If you read the linked issue, it's a hardcoded maximum recursion value in gRPC code so there isn't much we can do about it. Shouldn't affect the soon to be released python client v3 since it'll use the new encoding (a flat list).

**Comment by zurk** [comment]
I confirm that the issue is fixed in python client v3.0.4
Thanks!
